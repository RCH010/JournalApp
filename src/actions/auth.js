import { googleAuthProvider, firebase } from "../firebase/firebaseConfig";
import { types } from "../types/types"
import { finishLoading, startLoading } from "./ui";

//  CUANDO ES UNA TAREA ASINCRONA, SE NECESITA RETORNAR UN CALLBACK
export const startLoginWithEmailPassword = (email, password) => {
    return (dispatch) => {
        dispatch(startLoading());
        firebase.auth().signInWithEmailAndPassword(email,password)
            .then(({user}) => {
                dispatch(
                    login(user.uid, user.displayName, user.email)
                );
            })
            .catch(e => {
                console.log(e);
            })
            .finally(() => {
                dispatch(finishLoading())
            })
    }
}

export const startRegisterWithEmailPasswordName = (email, password, name) => {
    return (dispatch) => {
        firebase.auth().createUserWithEmailAndPassword(email,password)
            .then( async({user}) => {
                await user.updateProfile({displayName: name});
                dispatch(
                    login(user.uid, user.displayName, user.email)
                )
            })
            .catch( e => {
                console.log(e);
            })
    }
}

export const startGoogleLogin = () => {
    return (dispatch) => {
        firebase.auth().signInWithPopup(googleAuthProvider)
            .then(({user}) => {
                dispatch(
                    login(user.uid, user.displayName, user.email)
                )
            });
    }
}

export const login = (uid, displayName, email) => ( {
    type: types.login,
    payload: {
        uid,
        displayName,
        email
    }
})