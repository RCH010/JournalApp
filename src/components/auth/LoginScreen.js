import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { Link } from 'react-router-dom'
import validator from 'validator'
import { startGoogleLogin, startLoginWithEmailPassword } from '../../actions/auth'
import { removeError, setError } from '../../actions/ui'
import { useForm } from '../../hooks/useForm'

export const LoginScreen = () => {
    
    const dispatch = useDispatch()
    const {msgError, loading} = useSelector(state => state.ui);

    const [formValues, handleInputChange] = useForm({
        email: '',
        password: '',
    })

    const { email, password } = formValues;

    const handleLogin = (e) => {
        e.preventDefault();
        if(isFormValid()){
            dispatch(startLoginWithEmailPassword(email, password));
        }

    }

    const handleGoogleLogin = () => {
        dispatch(startGoogleLogin());
    }

    const isFormValid = () => {
        if(!validator.isEmail(email)) {
            dispatch(setError('Email is invalid.'));
            return false;
        }
        if(password.length <= 5) {
            dispatch(setError('Password should be at least 6 characters.'));
            return false;
        }
        dispatch(removeError());
        return true;
    }

    return (
        <div>
            <h3 className='auth__title'>Login</h3>

            <form onSubmit={handleLogin}>
                {
                    msgError && 
                    (<div className='auth__alert-error'>
                        {msgError}
                    </div>)
                }
                <input 
                    type='email'
                    placeholder='myjournal@gmail.com'
                    name='email'
                    className='auth__input'
                    autoComplete='off'
                    value={email}
                    onChange={handleInputChange}
                />
                <input 
                    type='password'
                    placeholder='Password'
                    name='password'
                    className='auth__input'
                    value={password}
                    onChange={handleInputChange}
                />

                <button
                    type='submit'
                    className='btn btn-primary btn-block'
                    disabled={loading}
                >
                    Login
                </button>
                

                <div className='auth__social-networks'>
                    <p className='auth__social-networks-title mt-1'>Login with social networks</p>
                
                    <button type="button" className="google-button mt-2" onClick={handleGoogleLogin}>
                        <span className="google-button__icon">
                            <svg viewBox="0 0 366 372" xmlns="http://www.w3.org/2000/svg"><path d="M125.9 10.2c40.2-13.9 85.3-13.6 125.3 1.1 22.2 8.2 42.5 21 59.9 37.1-5.8 6.3-12.1 12.2-18.1 18.3l-34.2 34.2c-11.3-10.8-25.1-19-40.1-23.6-17.6-5.3-36.6-6.1-54.6-2.2-21 4.5-40.5 15.5-55.6 30.9-12.2 12.3-21.4 27.5-27 43.9-20.3-15.8-40.6-31.5-61-47.3 21.5-43 60.1-76.9 105.4-92.4z" id="Shape" fill="#EA4335"/><path d="M20.6 102.4c20.3 15.8 40.6 31.5 61 47.3-8 23.3-8 49.2 0 72.4-20.3 15.8-40.6 31.6-60.9 47.3C1.9 232.7-3.8 189.6 4.4 149.2c3.3-16.2 8.7-32 16.2-46.8z" id="Shape" fill="#FBBC05"/><path d="M361.7 151.1c5.8 32.7 4.5 66.8-4.7 98.8-8.5 29.3-24.6 56.5-47.1 77.2l-59.1-45.9c19.5-13.1 33.3-34.3 37.2-57.5H186.6c.1-24.2.1-48.4.1-72.6h175z" id="Shape" fill="#4285F4"/><path d="M81.4 222.2c7.8 22.9 22.8 43.2 42.6 57.1 12.4 8.7 26.6 14.9 41.4 17.9 14.6 3 29.7 2.6 44.4.1 14.6-2.6 28.7-7.9 41-16.2l59.1 45.9c-21.3 19.7-48 33.1-76.2 39.6-31.2 7.1-64.2 7.3-95.2-1-24.6-6.5-47.7-18.2-67.6-34.1-20.9-16.6-38.3-38-50.4-62 20.3-15.7 40.6-31.5 60.9-47.3z" fill="#34A853"/></svg>
                        </span>
                        <span className="google-button__text">Sign in with Google</span>
                    </button>
                </div>
                <Link 
                    to='/auth/register'
                    className='link auth__new-account'
                >
                    Create new account
                </Link>
            </form>
        </div>
    )
}
