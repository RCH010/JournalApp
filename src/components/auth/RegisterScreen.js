import React from 'react'
import { Link } from "react-router-dom";
import { useForm } from '../../hooks/useForm';
import { useDispatch, useSelector } from 'react-redux';
import { removeError, setError } from '../../actions/ui';
import validator from 'validator';
import { startRegisterWithEmailPasswordName } from '../../actions/auth';

export const RegisterScreen = () => {

    const dispatch = useDispatch()
    const {msgError} = useSelector( state => state.ui);

    const [formValues, handleInputChange] = useForm({
        name: '',
        email: '',
        password: '',
        password2: '',
    });

    const { name, email, password, password2 } = formValues;

    const handleRegister = (e) => {
        e.preventDefault();
        if(isFormValid()) {
            dispatch(startRegisterWithEmailPasswordName(email,password,name));
        } 
    }

    const isFormValid = () => {
        if(name.trim().length === 0){
            dispatch(setError('Name is required.'));
            return false;
        }
        if (!validator.isEmail(email)){
            dispatch(setError('Email is not valid.'));
            return false;
        }
        if (password !== password2 || password.length <= 5){
            dispatch(setError('Password should be at least 6 characters and match with each other.'));
            return false;
        }
        dispatch(removeError());
        return true;
    }

    

    return (
        <div>
            <h3 className='auth__title'>Register</h3>

            <form onSubmit={handleRegister} className='animate__animated animate__fadeIn animate__faster'>
                {
                    msgError && 
                    (<div className='auth__alert-error'>
                        {msgError}
                    </div>)
                }
                <input 
                    type='text'
                    placeholder='Name'
                    name='name'
                    className='auth__input'
                    autoComplete='off'
                    value={name}
                    onChange={handleInputChange}
                />
                <input 
                    type='email'
                    placeholder='myjournal@gmail.com'
                    name='email'
                    className='auth__input'
                    autoComplete='off'
                    value={email}
                    onChange={handleInputChange}
                />
                <input 
                    type='password'
                    placeholder='Password'
                    name='password'
                    className='auth__input'
                    value={password}
                    onChange={handleInputChange}
                />
                <input 
                    type='password'
                    placeholder='Confirm Password'
                    name='password2'
                    className='auth__input'
                    value={password2}
                    onChange={handleInputChange}
                />

                <button
                    type='submit'
                    className='btn btn-primary btn-block mb-1 mt-1'
                >
                    Register
                </button>
                
                <Link 
                    to='/auth/login'
                    className='link'
                >
                    Already registered?
                </Link>
            </form>
        </div>
    )
}
