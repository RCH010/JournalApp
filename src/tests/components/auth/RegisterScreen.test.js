import { mount } from "enzyme";
import { act } from "react-dom/test-utils";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import configureStore from 'redux-mock-store';
import thunk from "redux-thunk";
import { RegisterScreen } from "../../../components/auth/RegisterScreen";
import { types } from "../../../types/types";

const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const initState = {
    auth: {},
    ui: {
        loading: false,
        msgError: null
    },
    notes: {
        notes: [],
        active: null
    }
};
let store = mockStore(initState);

describe('Tests in <RegisterScreen/>', () => {
    const wrapper = mount(
        <Provider store={store}>
            <MemoryRouter>
                <RegisterScreen />
            </MemoryRouter>
        </Provider>
    );

    beforeEach(()=> {
        store = mockStore(initState);
        jest.clearAllMocks();
    })

    test('should match with snapshot', () => {
        expect(wrapper).toMatchSnapshot();        
    })

    test('should dispatch error when fields are empty', () => {
        act(()=>{
            const wrapper = mount(
                <Provider store={store}>
                    <MemoryRouter>
                        <RegisterScreen />
                    </MemoryRouter>
                </Provider>
            );

            wrapper.find('form').prop('onSubmit')({ preventDefault(){} });
            const actions = store.getActions();
            expect(actions[0]).toEqual({
                type: types.uiSetError,
                payload: 'Name is required.'
            })
        })
    })

    test('should dispatch error if email is invalid', () => {
        const wrapper = mount(
            <Provider store={store}>
                <MemoryRouter>
                    <RegisterScreen />
                </MemoryRouter>
            </Provider>
        );
        
        wrapper.find({name:'name'}).simulate('change', { target: {name: 'name', value: 'Testing'} });
        wrapper.find({name:'email'}).simulate('change', { target: {name: 'email', value: 'test@testing.c'} });
        wrapper.find({name:'password'}).simulate('change', { target: {name: 'password', value: '1234567'} });
        wrapper.find({name:'password'}).simulate('change', { target: {name: 'password2', value: '1234567'} });
        
        wrapper.find('form').prop('onSubmit')({ preventDefault(){} });
        const actions = store.getActions();

        expect(actions[0]).toEqual({
            type: types.uiSetError,
            payload: 'Email is not valid.'
        })
    })

    test('should dispatch error if passwords dont match', () => {
        const wrapper = mount(
            <Provider store={store}>
                <MemoryRouter>
                    <RegisterScreen />
                </MemoryRouter>
            </Provider>
        );
        
        wrapper.find({name:'name'}).simulate('change', { target: {name: 'name', value: 'Testing'} });
        wrapper.find({name:'email'}).simulate('change', { target: {name: 'email', value: 'test@testing.com'} });
        wrapper.find({name:'password'}).simulate('change', { target: {name: 'password', value: '1234567'} });
        wrapper.find({name:'password'}).simulate('change', { target: {name: 'password2', value: '123457'} });
        
        wrapper.find('form').prop('onSubmit')({ preventDefault(){} });
        const actions = store.getActions();

        expect(actions[0]).toEqual({
            type: types.uiSetError,
            payload: 'Password should be at least 6 characters and match with each other.'
        })
    })
    
    test('should initiate with error', () => {
        const initState = {
            auth: {},
            ui: {
                loading: false,
                msgError: 'Email is not valid.'
            },
            notes: {
                notes: [],
                active: null
            }
        };
        store = mockStore(initState);
        const wrapper = mount(
            <Provider store={store}>
                <MemoryRouter>
                    <RegisterScreen />
                </MemoryRouter>
            </Provider>
        );
        expect(wrapper.find('.auth__alert-error').exists()).toBe(true);
    })

    test('should remove error when all is correct', () => {
        const initState = {
            auth: {},
            ui: {
                loading: false,
                msgError: 'Email is not valid.'
            },
            notes: {
                notes: [],
                active: null
            }
        };
        store = mockStore(initState);
        const wrapper = mount(
            <Provider store={store}>
                <MemoryRouter>
                    <RegisterScreen />
                </MemoryRouter>
            </Provider>
        );
        wrapper.find({name:'name'}).simulate('change', { target: {name: 'name', value: 'Testing'} });
        wrapper.find({name:'email'}).simulate('change', { target: {name: 'email', value: 'test@testing.com'} });
        wrapper.find({name:'password'}).simulate('change', { target: {name: 'password', value: '1234567'} });
        wrapper.find({name:'password'}).simulate('change', { target: {name: 'password2', value: '1234567'} });
        
        wrapper.find('form').prop('onSubmit')({ preventDefault(){} });
        const actions = store.getActions();
        expect(actions[0]).toEqual({
            type: types.uiRemoveError
        });
    })
})
