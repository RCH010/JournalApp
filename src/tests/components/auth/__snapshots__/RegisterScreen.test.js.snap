// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in <RegisterScreen/> should match with snapshot 1`] = `
<div>
  <h3
    className="auth__title"
  >
    Register
  </h3>
  <form
    className="animate__animated animate__fadeIn animate__faster"
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="auth__input"
      name="name"
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value=""
    />
    <input
      autoComplete="off"
      className="auth__input"
      name="email"
      onChange={[Function]}
      placeholder="myjournal@gmail.com"
      type="email"
      value=""
    />
    <input
      className="auth__input"
      name="password"
      onChange={[Function]}
      placeholder="Password"
      type="password"
      value=""
    />
    <input
      className="auth__input"
      name="password2"
      onChange={[Function]}
      placeholder="Confirm Password"
      type="password"
      value=""
    />
    <button
      className="btn btn-primary btn-block mb-1 mt-1"
      type="submit"
    >
      Register
    </button>
    <a
      className="link"
      href="/auth/login"
      onClick={[Function]}
    >
      Already registered?
    </a>
  </form>
</div>
`;
